### DISCUSSION TITLE
Reading from SHT4x (I2C) with ESP32

### DISCUSSION THREAD

**Discussion by affsantos** [micropython/micropython#9844] [Libraries & Drivers]
Hello,

I would like to read the temperature and humidity from SHT41 sensor with a ESP32 using I2C.

So far, I can see the sensor using i2c.scan() with the correct address (0x44). I can also write to the sensor without errors (using `i2c.writeto()`) but as soon I try to read from the sensor (using `i2c.readfrom()`) then I get the following error:

`OSError: [Errno 19] ENODEV`

Here is the code I am using (it fails when running the i2c.readfrom()):

```
    import machine
    import time

    i2c = machine.I2C(scl=machine.Pin(5), sda=machine.Pin(4))

    print('Scan i2c bus...')
    devices = i2c.scan()

    if len(devices) == 0:
      print("No i2c device !")
    else:
      print('i2c devices found:',len(devices))

      for device in devices:  
        print("Decimal address: ",device," | Hexa address: ",hex(device))
        
        
    SHT4x_ADDRESS = 0x44
    SHT4X_MEASURE = bytearray('\xFD')  

    i2c.writeto(SHT4x_ADDRESS, SHT4X_MEASURE)
    time.sleep_ms(10)
    unit = i2c.readfrom(SHT4x_ADDRESS, 6)
```

This code returns:
```
Scan i2c bus...
i2c devices found: 1
Decimal address:  68  | Hexa address:  0x44
Traceback (most recent call last):
File "<stdin>", line 23, in main
OSError: [Errno 19] ENODEV
```

Link to sensor datasheet: [https://download.mikroe.com/documents/d ... asheet.pdf](https://download.mikroe.com/documents/datasheets/SHT40%20Datasheet.pdf)

Note: [I found another person, in the old forum, having a similar issue but I could not find a solution for this problem](https://forum.micropython.org/viewtopic.php?t=12235).

What I tried:
- Different sensor (I have 2xSHT41, both same error)
- Different ESP32 and also tried with an old ESP8266, same error
- Add a pullup 1k resistor (same error) and 10k resistor (can't find the sensor anymore)
- Tried both I2C and SoftI2C libraries, same error.

Thanks in advance

**Comment by jimmo** [comment]
Just reading the [Adafruit driver for this sensor](https://github.com/adafruit/Adafruit_CircuitPython_SHT4x/blob/main/adafruit_sht4x.py) it looks like you're doing the right thing. The only thing I can see different is they trigger a soft reset by sending b'\x94' before the first read.

**Comment by jimmo** [comment]
Ahhh no I see it -- `bytearray('\xFD')` is not what you want. This is actually the bytes `0xc3`, `0xbd` which is the utf-8 representation of `0xfd`.

You need to use bytearray(b'\xFD')  (note the `b` prefix), but even better you don't actually need the bytearray at all -- just write `SHT4X_MEASURE = b'\xFD'`.

**Reply by jimmo** [reply]
CPython doesn't allow you to create a bytearray from a string like this, so I raised https://github.com/micropython/micropython/pull/9863 to try and avoid this trap!

**Reply by affsantos** [reply]
Thank you @jimmo for the help. I will keep an eye on your PR [#9863](https://github.com/micropython/micropython/pull/9863)!

**Reply by jimmo** [reply]
You don't need that PR though, that's just to prevent someone else running into the same surprise.

If you change your code as suggested then it should work.

**Reply by affsantos** [reply]
Thanks for the reply @jimmo.
I tried both `SHT4X_MEASURE = bytearray(b'\xFD')` and `SHT4X_MEASURE = b'\xFD'` but I got exactly the same error as before.

I also tried to implement the soft reset that you mentioned but I still get the same error. Below the code that I tried with the soft reset:

```
    SHT4X_RESET = b'\x94'
    SHT4x_ADDRESS = 0x44
    SHT4X_MEASURE = b'\xFD'
    
    i2c.writeto(SHT4x_ADDRESS, SHT4X_RESET)
    time.sleep_ms(10)
    i2c.writeto(SHT4x_ADDRESS, SHT4X_MEASURE)
    time.sleep_ms(10)
    unit = i2c.readfrom(SHT4x_ADDRESS, 6)
```
